provider/benchmark.go

```diff
@@ -0,0 +1,273 @@
+	t.Run("benchmark", func(t *testing.T) {
+		if !(BenchmarkFlag.IsSet() || RecordBenchmarkFlag.IsSet()) {
+			t.Skipf("Skip benchmarking because %s", BenchmarkFlag.WhyNotSet())
+		}
+
+		if bench.TracingDir == "" {
+			bench.TracingDir = t.TempDir()
+		}
+
+		// sanity check filesystem
+		ensureFolderExists(t, bench.TracingDir)
+		ensureFolderExists(t, bench.RecordDir)
+		_, err := os.Stat(bench.ProgramDir)
+		require.Nil(t, err)
+
+		// Collect traces
+		programTestAsSpanBenchmark(t, bench)
+
+		// After collecting traces, either compare or record
+		if RecordBenchmarkFlag.IsSet() {
+			copyFile(
+				t,
+				path.Join(bench.TracingDir, bench.outFileName()),
+				path.Join(bench.RecordDir, bench.outFileName()),
+			)
+			require.Nil(t, err)
+		} else {
+			// TODO(pulumi/home#3171)
+			t.Skip("TODO: implement comparison against recorded benchmarks")
+		}
+	})
+}
+
+// The core of trhe
+func programTestAsSpanBenchmark(t *testing.T, bench BenchmarkConfig) {
+
+	// Run preview only to make sure all needed plugins are
+	// downloaded so that these downloads do not skew
+	// measurements.
+	t.Run("prewarm", func(t *testing.T) {
+		prewarmOptions := bench.PrewarmProgramTestOptions()
+		integration.ProgramTest(t, &prewarmOptions)
+	})
+
+	benchFileName := path.Join(bench.TracingDir, bench.outFileName())
+	benchfile, err := os.OpenFile(benchFileName, os.O_WRONLY|os.O_CREATE, 0600)
+	require.Nil(t, err)
+	defer benchfile.Close()
+
+	for i, _ := range make([]struct{}, bench.Samples) {
+		t.Run(fmt.Sprintf("traced/%d", i), func(t *testing.T) {
+			opts := bench.TracingProgramTestOptions()
+			integration.ProgramTest(t, &opts)
+
+			err := TraceDirSpanIter{traceDir: bench.TracingDir}.WalkSpans(func(span Span) error {
+				benchLine, err := benchmarkFromSpan(span)
+				require.Nil(t, err)
+				if strings.Contains(benchLine, bench.Provider) {
+					benchfile.WriteString(benchLine)
+				}
+				return nil
+			})
+			require.Nil(t, err)
+		})
+	}
+}
+
+func ensureFolderExists(t *testing.T, dir string) {
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0755)
+		require.NoError(t, err)
+	}
+}
+
+func copyFile(t *testing.T, sourcePath, destPath string) {
+	inputFile, err := os.Open(sourcePath)
+	require.Nil(t, err)
+	defer inputFile.Close()
+
+	outputFile, err := os.Create(destPath)
+	require.Nil(t, err)
+	defer outputFile.Close()
+
+	_, err = io.Copy(outputFile, inputFile)
+	require.Nil(t, err)
+}
+
+// providertest doesn't expose a real flags implementation, so just pretend for now
+var BenchmarkFlag = hardcodedFlag{set: false}       //("benchmark", "Run benchmarks of provider test programs and compare with recorded values")
+var RecordBenchmarkFlag = hardcodedFlag{set: false} //("record-benchmark", "Run benchmarks of provider test programs and record results")
+
+type hardcodedFlag struct {
+	set bool
+}
+
+var _ = flags.Flag(hardcodedFlag{})
+
+func (h hardcodedFlag) IsSet() bool       { return h.set }
```

Comment:
Fancy stuff from providertest? We're getting a bit carried away there. I'm open to ripping it out; in the end a simple side-by-side test for the benchmark is not that bad. A new top level `func TestXYZPerf()`.